### PR TITLE
[Feature] Add shape-aware grouping head in SSN

### PR DESCRIPTION
**Pull request by Tai-Wang**
1. Add shape-aware grouping head in SSN (Shape Signature Networks)
2. Support different feature map sizes for grouping classes in anchor generator

### CODE REVIEW

File: mmdet3d/core/anchor/anchor_3d_generator.py [hunk 1]
```diff
@@ -323,3 +323,77 @@ def anchors_single_range(self,
...
         return ret
+
+
+@ANCHOR_GENERATORS.register_module()
+class AlignedAnchor3DRangeGeneratorPerCls(AlignedAnchor3DRangeGenerator):
+    """3D Anchor Generator by range for per class.
+
+    This anchor generator generates anchors by the given range for per class.
+    Note that feature maps of different classes may be different.
+    Args:
```
Comment (ZwwWayne) [inline mmdet3d/core/anchor/anchor_3d_generator.py:334]: Need an empty line.

File: mmdet3d/core/anchor/anchor_3d_generator.py [hunk 2]
```diff
@@ -323,3 +323,77 @@ def anchors_single_range(self,
...
+            ' not supported currently in this kind of anchor generator.'
+
+    def grid_anchors(self, featmap_sizes, device='cuda'):
+        """Generate grid anchors in multiple feature levels.
+
+        Args:
+            featmap_sizes (list[tuple]): List of feature map sizes for \
+                different classes in a single feature level.
+            device (str): Device where the anchors will be put on.
+        Returns:
```
Comment (ZwwWayne) [inline mmdet3d/core/anchor/anchor_3d_generator.py:351]: Empty line.

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+        conv_cfg (dict): Config of conv layer.
+            Default: dict(type='Conv2d')
+        norm_cfg (dict): Config of norm layer.
+            Default: dict(type='naiveSyncBN2d', eps=1e-3, momentum=0.01).
+        bias (bool|str): Type of bias. Default: False.
+    """
+
+    def __init__(self,
+                 num_cls,
+                 num_base_anchors,
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:35]: We may add an `num_shared_convs` to decide the number of shared convs, thus unifying small and large head.

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+
+from mmdet3d.core import box3d_multiclass_nms, limit_period, xywhr2xyxyr
+from mmdet.core import multi_apply
+from mmdet.models import HEADS
+from ..builder import build_head
+from .anchor3d_head import Anchor3DHead
+
+
+@HEADS.register_module()
+class BaseShapeHead(nn.Module):
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:14]: Considering reuse [BaseConvBboxHead](https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/models/dense_heads/base_conv_bbox_head.py#L9).

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+        for i, task in enumerate(self.tasks):
+            for _ in range(task['num_class']):
+                self.featmap_sizes.append(results[i]['featmap_size'])
+        assert len(self.featmap_sizes) == len(self.anchor_generator.ranges), \
+            'Length of feature map sizes must be equal to length of ' + \
+            'different ranges of anchor generator.'
+
+        return cls_score, bbox_pred, dir_cls_preds
+
+    def get_anchors(self, input_metas, device='cuda'):
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:368]: Do we need to rewrite the function `get_anchors'?

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+        """Get anchors according to feature map sizes.
+
+        Args:
+            input_metas (list[dict]): contain pcd and img's meta info.
+            device (str): device of current module.
+        Returns:
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:374]: Empty line.

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+    def loss(self,
+             cls_scores,
+             bbox_preds,
+             dir_cls_preds,
+             gt_bboxes,
+             gt_labels,
+             input_metas,
+             gt_bboxes_ignore=None):
+        """Calculate losses.
+        Args:
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:456]: Empty line

File: mmdet3d/models/dense_heads/shape_aware_head.py (hunk 5 above)
Comment (Tai-Wang) [reply mmdet3d/models/dense_heads/shape_aware_head.py]: Good idea. The previous get_anchors in the reformatting procedure can not be unified with the original one, but the current version allows. Already updated.

File: mmdet3d/models/dense_heads/shape_aware_head.py (hunk 3 above)
Comment (Tai-Wang) [reply mmdet3d/models/dense_heads/shape_aware_head.py:37]: Simplified and refactored.

File: mmdet3d/models/dense_heads/shape_aware_head.py (hunk 4 above)
Comment (Tai-Wang) [reply mmdet3d/models/dense_heads/shape_aware_head.py:14]: I would prefer to using current structure for consistency of the input names and meanings with the anchor3dhead. Reusing BaseConvBboxHead can not make it obviously simpler or clearer.

File: mmdet3d/core/anchor/anchor_3d_generator.py [hunk 8]
```diff
@@ -323,3 +323,79 @@ def anchors_single_range(self,
...
+        with different feature map sizes.
+
+        This function is usually called by method ``self.grid_anchors``.
+        Args:
+            featmap_sizes (list[tuple]): List of feature map sizes for \
+                different classes in a single feature level.
+            scale (float): Scale factor of the anchors in the current level.
+            device (str, optional): Device the tensor will be put on.
+                Defaults to 'cuda'.
+        Returns:
```
Comment (ZwwWayne) [inline mmdet3d/core/anchor/anchor_3d_generator.py:377]: Need an empty line above `Returns`, `Args`.

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+                 num_base_anchors,
+                 box_code_size,
+                 in_channels,
+                 shared_conv_channels=(64, 64, 64),
+                 shared_conv_strides=(2, 1, 1),
+                 use_direction_classifier=True,
+                 conv_cfg=dict(type='Conv2d'),
+                 norm_cfg=dict(type='naiveSyncBN2d', eps=1e-3, momentum=0.01),
+                 bias=False):
+        # Heavier head and smaller feature map than SmallHead
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:181]: May add this comment in the `Note` section in docstring.

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+                 num_base_anchors,
+                 box_code_size,
+                 in_channels,
+                 shared_conv_channels=(128, 128, 64, 64, 64),
+                 shared_conv_strides=(2, 1, 2, 1, 1),
+                 use_direction_classifier=True,
+                 conv_cfg=dict(type='Conv2d'),
+                 norm_cfg=dict(type='naiveSyncBN2d', eps=1e-3, momentum=0.01),
+                 bias=False):
+        # Heavier head and smaller feature map than SmallHead
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:223]: Move this into docstring. May also discuss the difference between it and the `LargeHead`.

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+            ]
+
+            input_meta = input_metas[img_id]
+            proposals = self.get_bboxes_single(cls_score_list, bbox_pred_list,
+                                               dir_cls_pred_list, mlvl_anchors,
+                                               input_meta, cfg, rescale)
+            result_list.append(proposals)
+        return result_list
+
+    def get_bboxes_single(self,
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:397]: This function seems to be duplicated, can we reuse the function implemented in the base class?

Comment (Tai-Wang) [reply mmdet3d/models/dense_heads/shape_aware_head.py:397]: It is not duplicated because the shapes of cls_score, bbox_preds, etc. are different from the implementation of anchor_3d_head.py (for example, [N, num_cls] vs. [num_cls, H, W]) due to different feature map sizes in SSN. So there are some minor differences when processing them in this functions.

**Comment by hellock** [comment]
Task linked: [CU-4hvduv [SSN] Support Shape-aware grouping head (week 1)](https://app.clickup.com/t/4hvduv)

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+        """Get bboxes of anchor head.
+
+        Args:
+            cls_scores (list[torch.Tensor]): Multi-level class scores.
+            bbox_preds (list[torch.Tensor]): Multi-level bbox predictions.
+            dir_cls_preds (list[torch.Tensor]): Multi-level direction
+                class predictions.
+            input_metas (list[dict]): Contain pcd and img's meta info.
+            cfg (None | :obj:`ConfigDict`): Training or testing config.
+            rescale (list[torch.Tensor]): Whether th rescale bbox.
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:361]: 1. (list[torch.Tensor]) -> (list[torch.Tensor], optional)
2. Default to False.

File: mmdet3d/models/dense_heads/shape_aware_head.py [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+                in single batch.
+            mlvl_anchors (List[torch.Tensor]): Multi-level anchors
+                in single batch.
+            input_meta (list[dict]): Contain pcd and img's meta info.
+            cfg (None | :obj:`ConfigDict`): Training or testing config.
+            rescale (list[torch.Tensor]): whether th rescale bbox.
+
+        Returns:
+            tuple: Contain predictions of single batch.
+                - bboxes (:obj:`BaseInstance3DBoxes`): Predicted 3d bboxes.
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/shape_aware_head.py:418]: Need an empty line here.

File: tests/test_heads.py [hunk 14]
```diff
@@ -945,3 +945,37 @@ def test_ssd3d_head():
...
+        'box_mode_3d': Box3DMode.LIDAR
+    }, {
+        'sample_idx': 2345,
+        'box_type_3d': LiDARInstance3DBoxes,
+        'box_mode_3d': Box3DMode.LIDAR
+    }]
+    (cls_score, bbox_pred, dir_cls_preds) = self.forward(feats)
+
+    # test get_boxes
+    cls_score[0] -= 1.5  # too many positive samples may cause cuda oom
```
Comment (encore-zhou) [inline tests/test_heads.py:1042]: Maybe we can modify the `score_thr` in `bbox_head_cfg` before calling `build_head`.

File: tests/test_heads.py [hunk 15]
```diff
@@ -945,3 +945,37 @@ def test_ssd3d_head():
...
+    result_list = self.get_bboxes(cls_score, bbox_pred, dir_cls_preds,
+                                  input_metas)
+    assert (result_list[0][1] > 0.3).all()
```
Comment (encore-zhou) [inline tests/test_heads.py:1045]: Maybe we need to check the size of `bboxes` firstly and also need to check the value of `labels`.

**Review by ZwwWayne — APPROVED** [review]
